TLS Support

patins opened the pull request:
Implements TLS Support for `PostgresBackend`. Feature now exposed in proxy, tests coming in a separate PR.
Feel free to review [rustls-split](https://github.com/patins/rustls-split) as well

TODO
- [x] lots of miscellaneous cleanup
- [x] force SSL
- [x] re-enable BufReader
- [x] PostgresBackend unit test

To address in separate PRs
- [ ] removed repeated code between pg_backend and proxy
- [ ] proxy integration tests (will do in a different PR)

hlinnaka commented on zenith_utils/src/pq_proto.rs:
```diff
@@ -346,7 +346,7 @@ pub enum BeMessage<'a> {
     ErrorResponse(String),
     // see https://www.postgresql.org/docs/devel/protocol-flow.html#id-1.10.5.7.11
-    Negotiate,
+    Negotiate(bool),
```
Looks like the PostgreSQL docs don't give a name to the `Y`/`N` response to an SSLRequest. `Negotiate` doesn't make it very clear what it is. I'd suggest `SSLHandshakeResponse` or something, with a comment pointing out it doesn't have a name in PostgreSQL.

I'm not sure how stable the link to the doc is, I think the anchors can change between PostgreSQL (major) versions. So in a year or two, that might point to wrong section.

hlinnaka commented on proxy/src/main.rs:
```diff
@@ -66,13 +106,16 @@ fn main() -> anyhow::Result<()> {
                 .help("redirect unauthenticated users to given uri")
                 .default_value("http://localhost:3000/psql_session/"),
         )
+        .arg(Arg::with_name("ssl-key"))
```
These really need a help-text. That's all the documentation we have for any of these options at the moment.

hlinnaka commented on proxy/src/proxy.rs:
```diff
@@ -226,31 +229,48 @@ databases without opening the browser.
 }
 
-async fn proxy_pass(pgb: PostgresBackend, db_info: DatabaseInfo) -> anyhow::Result<()> {
+async fn connect_to_db(db_info: DatabaseInfo) -> anyhow::Result<tokio::net::TcpStream> {
```
These functions desperately need comments.

hlinnaka commented on proxy/src/proxy.rs:
```diff
@@ -226,31 +229,48 @@ databases without opening the browser.
+    server_read: OwnedReadHalf,
+    server_write: OwnedWriteHalf,
+) -> anyhow::Result<()> {
+    fn do_proxy(mut reader: OwnedReadHalf, mut writer: OwnedWriteHalf) -> io::Result<()> {
```
Function in a function? Neat, I haven't seen that before :-)

hlinnaka reviewed (commented):
Hmm, how do we get about reviewing the `rustls-split` crate? We don't have a central place for general purpose crates like that, @ericseppanen also has the `aversion` and `bookfile` crates under his own github account. Move them all under the zenithdb org in github?

sharnoff commented on proxy/Cargo.toml:
```diff
@@ -17,5 +17,6 @@ serde_json = "1"
+rustls = "0.19.1"
 
-zenith_utils = { path = "../zenith_utils" }
+zenith_utils = { path = "../zenith_utils" }
```
missing a newline at the end of the file - not a big deal, but maybe worth fixing?

sharnoff commented on proxy/src/main.rs:
```diff
@@ -33,13 +36,57 @@ pub struct ProxyConf {
     pub waiters: Mutex<HashMap<String, mpsc::Sender<anyhow::Result<DatabaseInfo>>>>,
 }
 
+fn configure_ssl(arg_matches: &ArgMatches) -> anyhow::Result<Option<Arc<rustls::ServerConfig>>> {
```
This function is pretty visually messy. A couple suggestions below. Feel free to use them or leave them as you see fit :)

sharnoff commented on proxy/src/main.rs:
```diff
@@ -33,13 +36,57 @@ pub struct ProxyConf {
 }
 
+fn configure_ssl(arg_matches: &ArgMatches) -> anyhow::Result<Option<Arc<rustls::ServerConfig>>> {
+    match (
+        arg_matches.value_of("ssl-key"),
+        arg_matches.value_of("ssl-cert"),
+    ) {
```
Style alternative - completely up to you: when I get big match bodies like in this one, indentation can sometimes be a concern, so I'll often write something like this:

```rs
let (ssl_key_path, ssl_cert_path) = match (
    arg_matches.value_of("ssl-key"),
    arg_matches.value_of("ssl-cert"),
) {
    (Some(key), Some(cert)) => (key, cert),
    (None, None) => return Ok(None),
    _ => /* other error handling */,
};

/* Rest of existing match arm */
```

I think right now it's fine either way, but it's something that jumped out at me - thought I should mention.

sharnoff commented on proxy/src/main.rs:
```diff
@@ -33,13 +36,57 @@ pub struct ProxyConf {
+            let key = {
+                let key_bytes = std::fs::read(ssl_key_path).context("SSL key file")?;
+                let mut keys =
+                    rustls::internal::pemfile::rsa_private_keys(&mut Cursor::new(&key_bytes))
+                        .or_else(|_| {
+                            rustls::internal::pemfile::pkcs8_private_keys(&mut Cursor::new(
```
The `&mut Cursor::new(...)` stuff is a little confusing to understand why . Perhaps add a comment explaining that we need that there because the pemfile functions are expecting a `&mut dyn BufRead`?

patins replied:
I ended up switching to passing in `&mut &buf[..]`, as `&[u8]` implements BufRead as well

sharnoff commented on proxy/src/main.rs:
```diff
@@ -33,13 +36,57 @@ pub struct ProxyConf {
+            let key = {
+                let key_bytes = std::fs::read(ssl_key_path).context("SSL key file")?;
+                let mut keys =
+                    rustls::internal::pemfile::rsa_private_keys(&mut Cursor::new(&key_bytes))
+                        .or_else(|_| {
+                            rustls::internal::pemfile::pkcs8_private_keys(&mut Cursor::new(
+                                &key_bytes,
+                            ))
+                        })
```
There's a few of these really long paths that make it a bit harder to see what's going on. I'd add some `use` statements just inside the top of the function -- something like this might do the trick:
```rs
use rustls::internal::pemfile;
use rustls::{ServerConfig, NoClientAuth, ProtocolVersion::TLSv1_3};
use std::fs;
```

sharnoff commented on zenith_utils/src/postgres_backend.rs:
```diff
@@ -76,19 +79,48 @@ pub enum ProcessMsgResult {
     Break,
 }
 
+pub enum Stream {
```
It'd be nice to have a brief comment explaining why we need this (and perhaps what it is, though that's easier to understand from viewing it.)

sharnoff commented on zenith_utils/src/sock_split.rs:
```diff
@@ -0,0 +1,206 @@
+    }
+}
+
+pub enum ReadHalf {
```
IMO `ReadHalf`/`WriteHalf` might better named as `ReadStream`/`WriteStream` to match `BiDiStream`. I think "half" makes sense in the context of "splitting a socket", but it seems like most of the usage of this type isn't really with that context entirely present. Basically "half" seems like it'd be unclear unless you already know it's referring to a tcp stream.

Could also be something like `Tcp{Read,Write,Bidi}Stream` -- whatever you see fit. (Or not- just some thoughts)

sharnoff commented on zenith_utils/src/sock_split.rs:
```diff
@@ -0,0 +1,206 @@
+    }
+}
+
+pub enum BiDiStream {
```
Pointing it out here, but also occurs with `Stream::BiDirectional` in 'zenith_utils/src/postgres_backend'. Should it be "BiDi" or "Bidi"? Bidirectional is a single word, so I would have assumed it'd be `Bidi` (more so with `Stream::Bidirectional`). Wondering if you had a distinct reason for this

sharnoff reviewed (approved):
Had a look over: On the whole, I don't have a whole lot of substance to say about it. The general implementation strategy makes a bunch of sense. Adding an abstraction to split up streams seems like the right call to me, and aside from that the rest of it seems pretty straightforward.

Most of the things I commented on here are variations of "consider this style" or "perhaps a different name" - if you feel strongly about them, by all means leave it be.

With regards to `rustls-split`, I'd add some comments clarifying that the shared mutex is only ever shared between a reading half & writing half, which is why we don't end up with message interleaving. This is perhaps overly specific, but: reading the file from top to bottom meant I read the implementation of `io::Read` for `ReadHalf` (which drops the mutex guard before reading from the stream) before it was clear that the `Shared` component only had one `ReadHalf` viewing it, so there couldn't be racy behavior with concurrent calls to `read`. 

Overall, looks good. The comments in rustls-split aren't urgent (I'm happy to add them myself, if you'd like), and as long as you consider the things I've commented on, I think it's ok

patins commented:
Thanks for taking a look @sharnoff

I have cleaned up proxy and renamed the structs in `zenith_utils

`rustls-split` could certainly use some more comments. It also desperately needs docs to become a useful crate. I'll get on that
